batch/batch/driver/instance_collection/job_private.py

```diff
@@ -402,7 +425,12 @@ async def create_instance_with_error_handling(
                         batch_format_version = BatchFormatVersion(record['format_version'])
                         spec = json.loads(record['spec'])
                         machine_spec = batch_format_version.get_spec_machine_spec(spec)
-                        instance, total_resources_on_instance = await self.create_instance(machine_spec)
+
+                        regions = json_to_value(record['regions'])
```

Comment:
(This seems very reasonable)

Reply:
Not sure what you mean. The interface is that the user can specify `None` to mean any region. That is converted to no rows in `job_regions`. When we do the aggregation from the database, the returned value is `None`. We immediately convert that value to the list of all possible regions. From here on out, `regions` is no longer optional.

Reply:
Yeah we agree on the interpretation here. I'm saying that the user's interpretation of the "regions" of a job is this:

The "regions" of a job is the set of regions in which the job is allowed to run (represented as a list). You're allowed to say `[]` (i.e. the empty set) meaning no regions are allowed, but that's kind of silly (we should probably ban that in the front end). As a convenience for the user, instead of enumerating all the regions in GCP, we let them write `None` as a stand in for "the set of all regions".

The term "top" is used in lattice/order theory as a synonym for "maximum" or "greatest element". There's also "bottom" which has the opposite interpretation.

Reply:
We might even document that the `regions` of a job must be an element of the power set of the set of regions.

Reply:
Although I'm pretty comfortable with the idea of `None` filling this role (because I'm thinking of None as the unique extra element available in `Optional[X]`), I can also see how users would find this confusing (because "None" sounds like it means "no region"). It seems very reasonable to provide something like:

```
ANY_REGION = object()
```

In hailtop.batch.